### PR TITLE
CUMULUS-2107 add update-cmr-access-constraints task

### PR DESCRIPTION
**Pull request by nemreid**
**Summary:** Summary of changes

Addresses [CUMULUS-2107: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2107)

## Changes

* Add task for setting access constraints in CMR metadata.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [x] Integration tests

### CODE REVIEW

File: tasks/update-cmr-access-constraints/index.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+function updateGranuleCmrFileObjects(originalFiles, updatedFiles) {
+  updatedFiles.forEach((updatedFile) => {
+    originalFiles.filter(isCMRFile).forEach((file) => {
+      if (isSameCmrFileObject(file, updatedFile)) {
+        // eslint-disable-next-line no-param-reassign
+        file.etag = updatedFile.etag;
+      }
+    });
+  });
```
Comment (nemreid) [inline tasks/update-cmr-access-constraints/index.js:39]: This algorithm can almost certainly be optimized but I'm not quite sure how.

Comment (nemreid) [reply tasks/update-cmr-access-constraints/index.js]: Updated to a similar implementation.

File: tasks/update-cmr-access-constraints/index.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    cmrFileName,
+    cmrFileObject.etag
+  );
+  const { Bucket, Key } = parseS3Uri(cmrFileName);
+  if (isECHO10File(cmrFileName)) {
+    const updatedCmrMetadataContentsObject = setRestrictionMetadataInEcho10XMLMetadata(
+      cmrMetadataContentsObject,
+      accessConstraintsObject
+    );
+    const updatedGranuleXML = generateEcho10XMLString(updatedCmrMetadataContentsObject.Granule);
```
Comment (markdboyd) [inline tasks/update-cmr-access-constraints/index.js:86]: Weird, so we only store the contents of `Granule` from the XML?

File: tasks/update-cmr-access-constraints/tests/test-index.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  // ensure AccessConstraint is set in the metadata
+  const updatedMetadata = await metadataObjectFromCMRFile(
+    S3.buildS3Uri(t.context.bucket, key),
+    newETag
+  );
+  t.is(updatedMetadata.AccessConstraints.Value, accessConstraints.value);
+  t.is(updatedMetadata.AccessConstraints.Description, accessConstraints.description);
+});
+
+test('updateCmrAccessConstraints sets access constraint description to "None" if undefined', async (t) => {
```
Comment (markdboyd) [inline tasks/update-cmr-access-constraints/tests/test-index.js:102]: Technically this is done separately in the code for ECHO10 vs UMM-G, so might good to unit test it for both cases

File: tasks/update-cmr-access-constraints/schemas/input.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{
+  "title": "PostToCmrInput",
```
Comment (markdboyd) [inline tasks/update-cmr-access-constraints/schemas/input.json:2]: Should update the title/description here

File: tasks/update-cmr-access-constraints/schemas/output.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
+{
+  "title": "PostToCmrOutput",
```
Comment (markdboyd) [inline tasks/update-cmr-access-constraints/schemas/output.json:2]: Should update the title and description here

File: tasks/update-cmr-access-constraints/schemas/config.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  "title": "UpdateCmrAccessConstraintsConfig",
+  "description": "Describes the config used by the update-cmr-access-constraints task",
+  "type": "object",
+  "required": [
+    "accessConstraints"
+  ],
+  "properties": {
+    "accessConstraints": {
+      "type": "object",
+      "description": "object containing access constraint value and description",
```
Comment (markdboyd) [inline tasks/update-cmr-access-constraints/schemas/config.json:11]:
```suggestion
      "description": "object containing CMR access constraint value and description",
```

File: tasks/update-cmr-access-constraints/index.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    return { ...cmrFileObject, etag: updatedCmrFile.ETag };
+  }
+  throw new Error(`Unrecognized CMR File format: ${cmrFileName}`);
+}
+
+const updateCmrAccessConstraints = async (event) => {
+  const { config, input } = event;
+  const cmrFileObjects = granulesToCmrFileObjects(input.granules);
+  const updatedCmrFileObjectsWithEtags = await Promise.all(
+    cmrFileObjects.map((f) => updateCmrFileAccessConstraint(f, config.accessConstraints))
```
Comment (markdboyd) [inline tasks/update-cmr-access-constraints/index.js:111]:
```suggestion
    cmrFileObjects.map((cmrFileObject) => updateCmrFileAccessConstraint(cmrFileObject, config.accessConstraints))
```

File: tasks/update-cmr-access-constraints/index.js (hunk 2 above)
Comment (nemreid) [reply tasks/update-cmr-access-constraints/index.js:86]: Yes. It is actually the top level XML tag, but `metadataObjectFromCmrFile` returns it nested within an exterior JSON object.

File: tasks/update-cmr-access-constraints/index.js (hunk 7 above)
Comment (nemreid) [reply tasks/update-cmr-access-constraints/index.js]: fb86abf 👍

File: tasks/update-cmr-access-constraints/schemas/config.json (hunk 6 above)
Comment (nemreid) [reply tasks/update-cmr-access-constraints/schemas/config.json]: fb86abf 👍

File: tasks/update-cmr-access-constraints/schemas/output.json (hunk 5 above)
Comment (nemreid) [reply tasks/update-cmr-access-constraints/schemas/output.json]: fb86abf 👍

File: tasks/update-cmr-access-constraints/schemas/input.json (hunk 4 above)
Comment (nemreid) [reply tasks/update-cmr-access-constraints/schemas/input.json]: fb86abf 👍

File: tasks/update-cmr-access-constraints/tests/test-index.js (hunk 3 above)
Comment (nemreid) [reply tasks/update-cmr-access-constraints/tests/test-index.js]: fb86abf 👍

**Pull request merged**